File: java/src/main/java/com/bloock/quickstart/samples/encryption/Ecies.java
package com.bloock.quickstart.samples.encryption;

import com.bloock.quickstart.utils.Config;
import com.bloock.quickstart.utils.Logger;
import com.bloock.quickstart.utils.Sample;
import com.bloock.sdk.builder.Builder;
import com.bloock.sdk.client.Client;
import com.bloock.sdk.entity.EciesDecrypter;
import com.bloock.sdk.entity.EciesEncrypter;
import com.bloock.sdk.entity.KeyPair;
import com.bloock.sdk.entity.Record;

import java.nio.charset.StandardCharsets;

public class Ecies extends Sample {
  public void run(Config config) throws Exception {
    String payload = "This will be encrypted";
    Client sdk = new Client();

    KeyPair keypair = sdk.generateEciesKeyPair();

    Logger.info("The following payload will be encrypted: " + payload);

    // To encrypt we add an encrypter to the builder
    Record record = Builder.fromString(payload).withEncrypter(new EciesEncrypter(keypair.getPublicKey())).build();

    Logger.success("Encryption successful");

    Logger.success("Encrypted payload: " + new String(record.retrieve(), StandardCharsets.UTF_8));

    Logger.success("Record was created successfully");

    Logger.info("Trying to decrypt with the private key");

    // To decrypt we build a record from the encrypted record and add a decrypter
    Record decryptedRecord =
        Builder.fromRecord(record).withDecrypter(new EciesDecrypter(keypair.getPrivateKey())).build();

    Logger.success("Decryption successful");

    String hash = decryptedRecord.getHash();
    Logger.success("Hash: " + hash);

    Logger.success(
        "Decrypted payload: " + new String(decryptedRecord.retrieve(), StandardCharsets.UTF_8));

    assert decryptedRecord.retrieve().toString().equals(payload);

    Logger.info("Trying to decrypt with invalid key");

    Boolean exceptionWasThrown = false;
    try {
      Builder.fromRecord(record).withDecrypter(new EciesDecrypter("an invalid key")).build();
    } catch (Exception e) {
      exceptionWasThrown = true;
    }

    if (!exceptionWasThrown) {
      throw new Exception("The key was invalid so an error should've been returned!");
    }

    Logger.success("The key was invalid so the record could not be decrypted");
  }

  Ecies(String name) {
    super(name);
  }

  public static void main(String[] args) {
    new Ecies("ECIES Encryption");
  }
}

File: python/samples/encryption/ecies.py
from bloock.client.client import Client
from bloock.client.entity.decrypter import EciesDecrypter
from bloock.client.entity.encrypter import EciesEncrypter
from utils.logger import Logger
from utils.sample import Sample
from utils.config import Config

from bloock.client.builder import RecordBuilder


def ecies(_: Config):
    sdk = Client()

    payload = "This will be encrypted"

    keypair = sdk.generate_ecies_keypair()

    Logger.warn(f'The following payload will be encrypted: "{payload}"')

    # To encrypt we add an encrypter to the builder
    encrypted_record = (
        RecordBuilder.from_string(payload)
        .with_encrypter(EciesEncrypter(keypair.public_key))
        .build()
    )

    Logger.success("Encryption successful")
    Logger.success(f"Encrypted payload: {encrypted_record.retrieve().decode()}")

    Logger.warn("Trying to decrypt with the public key")

    # To decrypt we build a record from the encrypted record and add a decrypter
    decrypted_record = (
        RecordBuilder.from_record(encrypted_record)
        .with_decrypter(EciesDecrypter(keypair.private_key))
        .build()
    )

    Logger.success(f"Decryption successful")

    hash = decrypted_record.get_hash()
    Logger.success(f"Hash: {hash}")
    Logger.success(f'Decrypted payload: "{decrypted_record.retrieve().decode()}"')

    Logger.warn("Trying to decrypt with invalid private key")

    exception_was_thrown = False
    try:
        RecordBuilder.from_record(encrypted_record).with_decrypter(
            EciesDecrypter("an invalid key")
        ).build()
    except:
        exception_was_thrown = True

    if not exception_was_thrown:
        raise Exception("The key was invalid so an error should've been returned!")

    Logger.success("The key was invalid so the record could not be decrypted")


Sample("ECIES encryption", ecies)
